libcxx/include/__type_traits/is_trivially_relocatable.h

```diff
@@ -23,14 +23,12 @@ _LIBCPP_BEGIN_NAMESPACE_STD
 
 // A type is trivially relocatable if a move construct + destroy of the original object is equivalent to
 // `memcpy(dst, src, sizeof(T))`.
-
-#if __has_builtin(__is_trivially_relocatable)
-template <class _Tp, class = void>
-struct __libcpp_is_trivially_relocatable : integral_constant<bool, __is_trivially_relocatable(_Tp)> {};
-#else
+//
+// Note that we don't use Clang's __is_trivially_relocatable builtin because it doesn't honor the presence
+// of non-trivial special members like assignment operators, or even a copy constructor, making it possible
```

Comment:
In particular, this change is required or else we fail the test I added in `erase_iter.pass.cpp` which counts the number of assignments. Because of this, and since the semantics of what it means to be trivially relocatable are still being actively debated, I think it makes sense for us to err on the safer side and use `is_trivially_copyable` for now.

Reply:
`TrackedAssignment` has a trivial destructor and a trivial copy constructor, so it's trivially relocatable by destroying it and copy-constructing a new instance. Is the library code not permitted to do that? If not, would it make sense to have separate internal traits for "trivially assignment-relocatable" versus "trivially construction-relocatable" that only detect the case where you could relocate specifically using assignment versus specifically using construction?

Reply:
Sorry for the delay, the last days have been busy. I believe what you are suggesting is essentially what http://wg21.link/P2687R9 introduced with the `replaceable` concept -- or were you aware of that concept but you had something else in mind?

Since I wrote this patch, my understanding has evolved and so did P2687, so I created [this draft PR](https://github.com/llvm/llvm-project/pull/116714/commits) to discuss with the P2687 authors. That PR implements the currently proposed semantics for P2687 (although with a IMO better library API).